frame/crowdloan-rewards/src/lib.rs

```diff
@@ -423,3 +468,112 @@ pub mod pallet {
 		Some(addr)
 	}
 }
+
+/// Validate `associate` calls prior to execution. Needed to avoid a DoS attack since they are
+/// otherwise free to place on chain.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct PrevalidateAssociation<T: Config + Send + Sync>(sp_std::marker::PhantomData<T>)
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>;
+
+impl<T: Config + Send + Sync> sp_std::fmt::Debug for PrevalidateAssociation<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "PrevalidateAssociation")
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+#[allow(clippy::new_without_default)]
+impl<T: Config + Send + Sync> PrevalidateAssociation<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	/// Create new `SignedExtension` to validate crowdloan rewards association
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Config + Send + Sync> SignedExtension for PrevalidateAssociation<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	type AccountId = T::AccountId;
+	type Call = <T as frame_system::Config>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "PrevalidateAssociation";
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(())
+	}
+
+	// <weight>
+	// The weight of this logic is included in the `associate` dispatchable.
+	// </weight>
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		use frame_support::traits::Get;
+
+		if let Some(Call::associate { reward_account, proof }) = IsSubType::is_sub_type(call) {
+			if Associations::<T>::get(reward_account).is_some() {
+				return InvalidTransaction::Custom(ValidityError::AlreadyAssociated as u8).into()
+			}
+
+			let remote_account = match proof {
+				Proof::Ethereum(eth_proof) => {
+					let reward_account_encoded =
+						reward_account.using_encoded(|x| hex::encode(x).as_bytes().to_vec());
```

Comment:
this hex part is copy pased, can be reused

Reply:
not sure, if you are not going to take weight, than remove weight from assosiate call. amount of compute will be same until error.  i do not get difference how checking here safes from ddos? it is kind of optimization so the call does not gets into queue or something?

Reply:
As per the [docs](https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/traits.rs#L860-L877), I believe it keeps it out of the queue if it's not validated

Reply:
See latest commit with deduplicated code

Reply:
This should be fine as it is a signed extension. Meaning that if the check does not pass, the extrinsic is not pushed to the mempool, hence not propagated to the network